### DISCUSSION TITLE
[DI] How to setup Autowire(tags: [...]) + TaggedIterator + a chain pattern

### DISCUSSION THREAD

**Discussion by gnutix** [symfony/symfony#45312] [Q&A] [answered]
Hello folks,

I wanted to have the following classes for a 2FA model : 

```php
/*
 * Classes
 */

interface MethodInterface {}

final class SmsMethod implements MethodInterface
{
    // ...
}
final class TotpMethod implements MethodInterface
{
    // ...
}

/*
 * Factories
 */

#[Autoconfigure(tags: [MethodFactoryInterface::class])]
interface MethodFactoryInterface
{
    public function supports(string $uri): bool;

    public function createFromUri(string $uri): MethodInterface;
}

final class SmsMethodFactory implements MethodFactoryInterface
{
    public function supports(string $uri): bool
    {
        return str_starts_with($uri, 'sms:');
    }

    public function createFromUri(string $uri): MethodInterface
    {
        return new SmsMethod(...);
    }
}

final class TotpMethodFactory implements MethodFactoryInterface
{
    public function supports(string $uri): bool
    {
        return str_starts_with($uri, 'totp:');
    }

    public function createFromUri(string $uri): MethodInterface
    {
        return new TotpMethod(...);
    }
}

final class ChainMethodFactory implements MethodFactoryInterface
{
    private iterable $factories;

    public function __construct(
        #[TaggedIterator(MethodFactoryInterface::class) iterable $factories,
    ) {
        $this->factories = $factories;
    }

    public function supports(string $uri): bool
    {
        return true;
    }

    public function createFromUri(string $uri): MethodInterface
    {
        foreach ($this->factories as $factory) {
            if (!$factory->supports($uri)) {
                continue;
            }

            return $factory->createFromUri($uri);
        }

        throw new \Exception('No factory can deal with URI : ' . $uri);
    }
}

/*
 * Services
 */

final class SomeTwoFactorService
{
    private MethodFactoryInterface $methodFactory;

    public function __construct(
        #[Target(ChainMethodFactory::class)] MethodFactoryInterface $methodFactory;
    ) {
        $this->methodFactory = $methodFactory;
    }

    public function createSomeOtherObject(string $uri): SomeOtherObject
    {
        return new SomeOtherObject(method: $this->methodFactory->createFromUri($uri));
    }
}

# services.php

$services->alias(MethodFactoryInterface::class, ChainMethodFactory::class);
```

As far as design goes "on paper", everything seemed okay to me : the `ChainMethodFactory` is a `MethodFactoryInterface` itself, that delegates to other `MethodFactoryInterface` implementations. However, there is one major problem : how to avoid an infinite loop, as `ChainMethodFactory` is passed to its own constructor through the `#[TaggedIterator]` attribute (as it implements `MethodFactoryInterface` too) ? (I get a `502 Bad Gateway`)

My first instinct was to set a different tag on `ChainMethodFactory` : `#[Autowire(tags: [ChainMethodFactory::class])`, but it did not work... I'm not sure how inheritance works exactly when using such PHP8 attributes (and the current documentation around them is clearly lacking), but I guess it just ended up having the two tags instead of one...

So in the end I found a workaround by declaring the `#[Autowire(tags: MethodFactoryInterface::class)]` attribute only on the "standard" implementations (`SmsMethodFactory`, `TotpMethodFactory`) (and not on `ChainMethodFactory`). But that means a develop that creates a new implementation must remember to add this attribute, otherwise the implementation won't get picked up by the `ChainMethodFactory`...

Any idea how to avoid that ? I've read about service subscribers / locators in the documentation, but I don't see how they would help in this situation, as they always seem to have a "static" list of services they require, and the whole point of my design is that any new class implementating `MethodFactoryInterface` gets properly used in the system automatically.

Thanks for your insights.
gnutix

**Comment by norkunas** [comment, marked as answer]
https://github.com/symfony/symfony/pull/44774 this will help in future :)

**Reply by gnutix** [reply]
Too bad it's not backported, the feature is kind of broken without this code...

**Reply by norkunas** [reply]
Yeah needed it too..so just configured things in services.yaml with exclude of chained service for tag

**Reply by ruudk** [reply]
> Too bad it's not backported, the feature is kind of broken without this code...

It cannot be back ported as it's a new feature, not a bug fix.
